### PR TITLE
Lambda authorizers support (API, DataStore, PubSub)

### PR DESCRIPTION
**Pull request by iartemiev**
note: opening this as a branch in upstream in order to run e2e tests. 

#### Description of changes

Adds support for Lambda Authorizers for AppSync APIs to the API category and DataStore.

* refactored/DRY'd some of our existing unit tests

### API Category

```js
let authToken = 'some-auth-token-string'; // developer-managed token

// if AWS_LAMBDA is the default auth type, i.e., aws_appsync_authenticationType: 'AWS_LAMBDA':
const blogs = await API.graphql({query: queries.listBlogs, authToken: authToken});

// if you have a different default auth type configured:
const blogs = await API.graphql({query: queries.listBlogs, authMode: 'AWS_LAMBDA', authToken: authToken});
```

### DataStore

```js
let authToken = 'some-auth-token-string'; // developer-managed token

DataStore.configure({
  authProviders: {
    functionAuthProvider: () => (
      {
        token: authToken,
      }
    ),
  },
});

// Supports promises
DataStore.configure({
  authProviders: {
    functionAuthProvider: async () => {
      const authToken = await getAuthToken();
      return {
        token: authToken,
      }
    },
  },
});

const blogs = await DataStore.query(Blog);
```

#### Description of how you validated changes
* Added unit tests
* Manual QA
* All existing e2e tests pass: https://app.circleci.com/pipelines/github/aws-amplify/amplify-js/8436/workflows/d8e67470-87be-4e28-99c0-b4b13f43c91d
* New e2e tests added: https://github.com/aws-amplify/amplify-js-samples-staging/pull/247

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced) (will be added before release to latest)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/src/authModeStrategies/multiAuthStrategy.ts [hunk 1]
```diff
@@ -51,6 +57,11 @@ function getAuthRules({
 
 	rules.forEach(rule => {
 		switch (rule.allow) {
+			case 'custom':
+				if (rule.provider === 'function') {
```
Comment (nickarocho) [inline packages/datastore/src/authModeStrategies/multiAuthStrategy.ts:61]: Just curious about the other use cases for `rule.provider`? When would it not be a function?

**Review by nickarocho — APPROVED** [review]
LGTM! Super clean, great work @iartemiev 🥂 Would love to learn more about the use cases and can't wait to try this out once it's available.

File: packages/datastore/src/authModeStrategies/multiAuthStrategy.ts (hunk 1 above)
Comment (iartemiev) [reply packages/datastore/src/authModeStrategies/multiAuthStrategy.ts:61]: For this new `custom` auth strategy, `function` is the only valid provider at the moment, but we may add additional ones in the future. I added the `if` statement there to make it clear where to handle a new provider if/when we extend it down the road

Comment (nickarocho) [reply packages/datastore/src/authModeStrategies/multiAuthStrategy.ts:61]: Nice! Very scalable and forward-thinking, great stuff Ivan 😊🚀

**Review by svidgen — APPROVED** [review]
No blockers that I can see.

One thing I'd note overall, in please where we've created constants like `ModelAttributeAuthProvider.FUNCTION`, we really ought to be using them for the purposes of maintainability, semantic clarity, and fat-finger proofing.

**Comment by iartemiev** [comment]
Note: I'm going to do a normal PR merge instead of squashing in this case because I want API, PubSub, and DataStore to all get minor version bumps from this PR.

I've rebased/fixed up the previous commit messages into 3 commits total (one per category), so the commit history will remain clean and relevant.

**Pull request merged**